### PR TITLE
ruff: Build as a standalone without Python

### PR DESCRIPTION
**Pull request by adisbladis**
Using `buildPythonPackage` make `ruff` propagate the Python used for the build, which might not be the same Python as you want in your development shell.

https://github.com/NixOS/nixpkgs/pull/350654 changed the ruff top-level attribute to be built with Python modules. This doesn't actually make much sense, we have versioned Python sets, and including just the one Python in the top-level package isn't helpful, and causes issues downstream related to PATH & PYTHONPATH.

See my related [uv PR](https://github.com/NixOS/nixpkgs/pull/357113#issue-2670270577).

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by drupol — APPROVED** [review]
I agree with this.

**Review by baloo — CHANGES_REQUESTED** [review]
Just change your development shell expression to use `python312.pkgs.ruff` (or whichever version of python you might want) and not just `ruff`? 
I'm not sure I follow the problem here.

The upstream project intends for ruff to be built using maturin. Here you're duplicating the build to make another one that won't use maturin.

**Comment by baloo** [comment]
If the purpose is to strip off propagation, you can do so in your nix-shell:
```nix
let
  pkgs = import <nixpkgs> {};
  ruff = pkgs.runCommandNoCC "ruff" {} ''
    mkdir -p $out/bin
    ln -s ${pkgs.ruff}/bin/ruff $out/bin
  '';

in
  pkgs.mkShell {
    buildInputs = [ ruff  ];
}
```

```
$ nix-shell --command 'env| grep -i python'
$
```

**Comment by drupol** [comment]
> If the purpose is to strip off propagation, you can do so in your nix-shell

But this should be the default behaviour, don't you think?

**Comment by dwt** [comment]
@baloo Well, you have to start somewhere. Starting with popular packages that matter to you seems an apt strategy?

**Comment by baloo** [comment]
Well, the problem can be solved by using `ruff.bin` instead of `ruff` which doesn't bring in any python dependency or `PYTHON_PATH` (so you don't even need the `runCommand` wrapper like I suggested earlier).

**Review by drupol — APPROVED** [review]

File: pkgs/by-name/ru/ruff/package.nix [hunk 1]
```diff
@@ -39,38 +33,25 @@ python3Packages.buildPythonPackage rec {
...
-      mv $out/bin/ruff $bin/bin/
-      rmdir $out/bin
-    ''
-    + lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
-      installShellCompletion --cmd ruff \
-        --bash <($bin/bin/ruff generate-shell-completion bash) \
-        --fish <($bin/bin/ruff generate-shell-completion fish) \
-        --zsh <($bin/bin/ruff generate-shell-completion zsh)
-    '';
+  postInstall = lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
```
Comment (ambroisie) [inline pkgs/by-name/ru/ruff/package.nix:45]: Nit: could use `stdenv.hostPlatform.emulator buildPackages` instead.

File: pkgs/by-name/ru/ruff/package.nix [hunk 2]
```diff
@@ -127,7 +107,7 @@ python3Packages.buildPythonPackage rec {
   meta = {
     description = "Extremely fast Python linter";
     homepage = "https://github.com/astral-sh/ruff";
-    changelog = "https://github.com/astral-sh/ruff/releases/tag/${version}";
+    changelog = "https://github.com/astral-sh/ruff/releases/tag/${src.tag}";
```
Comment (ambroisie) [inline pkgs/by-name/ru/ruff/package.nix:110]: Shouldn't use `src` internals for `meta` (see #367739).

File: pkgs/by-name/ru/ruff/package.nix (hunk 1 above)
Comment (ambroisie) [reply pkgs/by-name/ru/ruff/package.nix]: Almost, you don't need `stdenv.buildPlatform.canExecute` as a condition for it. You can look at how other packages do it.

**Review by drupol — APPROVED** [review]

File: pkgs/development/python-modules/ruff/default.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
+  buildPythonPackage,
+  pkgs,
```
Comment (baloo) [inline pkgs/development/python-modules/ruff/default.nix:3]: This effectively breaks cross-compilation.

**Review by baloo — DISMISSED** [review]
This breaks `python3Packages.pytest-ruff` which currently builds on master.

I think it requires the following patch:
```
diff --git a/pkgs/by-name/ru/ruff/package.nix b/pkgs/by-name/ru/ruff/package.nix
index 646ef93836f5..662b7409ec8e 100644
--- a/pkgs/by-name/ru/ruff/package.nix
+++ b/pkgs/by-name/ru/ruff/package.nix
@@ -26,14 +26,6 @@ rustPlatform.buildRustPackage rec {
     hash = "sha256-c5d2XaoEjCHWMdjTLD6CnwP8rpSXTUrmKSs0QWQ6UG0=";
   };

-  # Do not rely on path lookup at runtime to find the ruff binary
-  postPatch = ''
-    substituteInPlace python/ruff/__main__.py \
-      --replace-fail \
-        'ruff_exe = "ruff" + sysconfig.get_config_var("EXE")' \
-        'return "${placeholder "bin"}/bin/ruff"'
-  '';
-
   useFetchCargoVendor = true;
   cargoHash = "sha256-jbUjsIJRpkKYc+qHN8tkcZrcjPTFJfdCsatezzdX4Ss=";

diff --git a/pkgs/development/python-modules/ruff/default.nix b/pkgs/development/python-modules/ruff/default.nix
index 8d79eee66f4b..8f5efa9a240b 100644
--- a/pkgs/development/python-modules/ruff/default.nix
+++ b/pkgs/development/python-modules/ruff/default.nix
@@ -10,12 +10,19 @@ buildPythonPackage {
     pname
     version
     src
-    postPatch
     cargoDeps
     postInstall
     meta
     ;

+  # Do not rely on path lookup at runtime to find the ruff binary
+  postPatch = ''
+    substituteInPlace python/ruff/__main__.py \
+      --replace-fail \
+        'ruff_exe = "ruff" + sysconfig.get_config_var("EXE")' \
+        'return "${placeholder "out"}/bin/ruff"'
+  '';
+
   pyproject = true;

   nativeBuildInputs = [
```

(the python derivation does not carry a `bin` output)

**Review by drupol — APPROVED** [review]

File: pkgs/development/python-modules/ruff/default.nix (hunk 3 above)
Comment (dotlambda) [reply pkgs/development/python-modules/ruff/default.nix:3]:
```suggestion
  ruff,
```

File: pkgs/development/python-modules/ruff/default.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  rustPlatform,
+  installShellFiles,
+}:
+
+buildPythonPackage {
+  inherit (pkgs.ruff)
```
Comment (dotlambda) [inline pkgs/development/python-modules/ruff/default.nix:9]:
```suggestion
  inherit (ruff)
```

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -14282,9 +14282,7 @@ self: super: with self; {
 
   rubymarshal = callPackage ../development/python-modules/rubymarshal { };
 
-  ruff = toPythonModule (pkgs.ruff.override {
-    python3Packages = self;
-  });
+  ruff = callPackage ../development/python-modules/ruff { };
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix:14285]:
```suggestion
  ruff = callPackage ../development/python-modules/ruff {
    inherit (pkgs) ruff;
  };
```

**Pull request merged**